neo/Network/P2P/Message.cs

```diff
@@ -1,79 +1,130 @@
+                }
+            }
+
             return new Message
             {
+                Flags = flags,
                 Command = command,
-                Checksum = GetChecksum(payload),
                 Payload = payload
             };
         }
 
+        void ISerializable.Serialize(BinaryWriter writer)
+        {
+            writer.Write((byte)Flags);
+            writer.Write((byte)Command);
+            writer.WriteVarBytes(Payload);
+        }
+
         void ISerializable.Deserialize(BinaryReader reader)
         {
-            if (reader.ReadUInt32() != Magic)
-                throw new FormatException();
-            this.Command = reader.ReadFixedString(12);
-            uint length = reader.ReadUInt32();
-            if (length > PayloadMaxSize)
-                throw new FormatException();
-            this.Checksum = reader.ReadUInt32();
-            this.Payload = reader.ReadBytes((int)length);
-            if (GetChecksum(Payload) != Checksum)
-                throw new FormatException();
+            Flags = (MessageFlags)reader.ReadByte();
+            Command = (MessageCommand)reader.ReadByte();
+            Payload = reader.ReadVarBytes(PayloadMaxSize);
         }
 
-        private static uint GetChecksum(byte[] value)
+        public static int TryDeserialize(ByteString data, out Message msg)
         {
-            return Crypto.Default.Hash256(value).ToUInt32(0);
+            msg = null;
+            if (data.Count < 3) return 0;
+
+            var header = data.Slice(0, 5).ToArray();
+            var flags = (MessageFlags)header[0];
+            ulong length = header[2];
+            var payloadIndex = 3;
+
+            if (length == 0xFD)
+            {
+                if (data.Count < 5) return 0;
+                length = data.Slice(payloadIndex, 2).ToArray().ToUInt16(0);
+                payloadIndex += 2;
+            }
+            else if (length == 0xFE)
+            {
+                if (data.Count < 7) return 0;
+                length = data.Slice(payloadIndex, 4).ToArray().ToUInt32(0);
+                payloadIndex += 4;
+            }
+            else if (length == 0xFF)
+            {
+                if (data.Count < 11) return 0;
+                length = data.Slice(payloadIndex, 8).ToArray().ToUInt64(0);
+                payloadIndex += 8;
+            }
+
+            if (length > PayloadMaxSize) throw new FormatException();
+
+            if (data.Count < (int)length + payloadIndex) return 0;
+
+            msg = new Message()
+            {
+                Flags = flags,
+                Command = (MessageCommand)header[1],
+                Payload = data.Slice(payloadIndex, (int)length).ToArray()
+            };
+
+            return payloadIndex + (int)length;
         }
 
-        private ISerializable _payload_deserialized = null;
+        public byte[] GetPayload() => Flags.HasFlag(MessageFlags.Compressed) ? Payload.DecompressLz4() : Payload;
```

Comment:
We'd better cache the decompressed payload.

Reply:
Because is cached the deserialized object, i think that this call will be called only once

Reply:
Then we need to deserialize the payload when deserializing the message. Because we need to limit the size of the decompressed payload to `PayloadMaxSize`.

Reply:
@shargon I created a PR: https://github.com/shargon/neo/pull/9